### PR TITLE
Adding TLS Client CA checks to diagnose

### PR DESCRIPTION
**Pull request by hghaf099**

### CODE REVIEW

File: vault/diagnose/tls_verification.go [hunk 1]
```diff
@@ -122,3 +123,30 @@ func TLSFileChecks(certFilePath, keyFilePath string) error {
...
+	// TODO: should we run the CA cert through the TLSFileChecks function as well?
+	return nil
+}
+
+// TLSMutualExclusionCertCheck returns error if both TLSDisableClientCerts and TLSRequireAndVerifyClientCert are set
+func TLSMutualExclusionCertCheck(l *configutil.Listener) error {
+
+	if l.TLSDisableClientCerts {
+		if l.TLSRequireAndVerifyClientCert {
+			return fmt.Errorf("'tls_disable_client_certs' and 'tls_require_and_verify_client_cert' are mutually exclusive")
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification.go:148]:
```suggestion
			return fmt.Errorf("the tls_disable_client_certs and tls_require_and_verify_client_cert fields in the listener stanza of the vault server config are mutually exclusive fields. Please ensure they are not both set to true.")
```

File: command/operator_diagnose.go [hunk 2]
```diff
@@ -489,7 +490,12 @@ SEALFAIL:
 				if err != nil {
 					return err
 				}
-
+				fmt.Println("Checking TLS Client CA File Check")
+				err = diagnose.TLSClientCAFileCheck(ln.Config)
+				fmt.Println("Checking TLS cleint CA file check and the error is : ", err)
```
Comment (HridoyRoy) [inline command/operator_diagnose.go:495]: Let's remove these print statements.

File: vault/diagnose/tls_verification.go [hunk 3]
```diff
@@ -122,3 +123,30 @@ func TLSFileChecks(certFilePath, keyFilePath string) error {
...
+
+	if !caPool.AppendCertsFromPEM(data) {
+		return fmt.Errorf("failed to parse CA certificate in tls_client_ca_file")
+	}
+	// TODO: should we run the CA cert through the TLSFileChecks function as well?
+	return nil
+}
+
+// TLSMutualExclusionCertCheck returns error if both TLSDisableClientCerts and TLSRequireAndVerifyClientCert are set
+func TLSMutualExclusionCertCheck(l *configutil.Listener) error {
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification.go:144]: This function needs to be called in `operator_diagnose`.

File: vault/diagnose/tls_verification.go [hunk 4]
```diff
@@ -122,3 +123,30 @@ func TLSFileChecks(certFilePath, keyFilePath string) error {
 
 	return nil
 }
+
+// TLSClientCAFileCheck checks whether client CA file is readable and parsable or not
+func TLSClientCAFileCheck(l *configutil.Listener) error {
+
+	caPool := x509.NewCertPool()
+	data, err := ioutil.ReadFile(l.TLSClientCAFile)
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification.go:131]: This works, but I think for the sake of reusing code, we should use https://github.com/hashicorp/vault/blob/13629ca2d4e8ddd221086751d9a09d75bd46cf49/vault/diagnose/tls_verification.go#L97

File: vault/diagnose/tls_verification_test.go [hunk 5]
```diff
@@ -315,3 +315,69 @@ func TestTLSInvalidMaxVersion(t *testing.T) {
 		t.Errorf("Bad error message: %w", err)
 	}
 }
+
+func TestTLSValidClientCAFile(t *testing.T) {
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification_test.go:319]: Could we add comments to these tests?

File: internalshared/listenerutil/listener.go [hunk 6]
```diff
@@ -163,6 +163,7 @@ Please see https://tools.ietf.org/html/rfc7540#appendix-A for further informatio
 
 	if l.TLSRequireAndVerifyClientCert {
 		tlsConf.ClientAuth = tls.RequireAndVerifyClientCert
+		// TODO: Wouldn't we have to provide the TLSClientCAFile if TLSRequireAndVerifyClientCert is set?
```
Comment (HridoyRoy) [inline internalshared/listenerutil/listener.go:166]: Can we remove this comment and track investigation for this in a jira? Thanks!

**Review by HridoyRoy — CHANGES_REQUESTED** [review]
A few quick formatting changes, but looking good!

File: vault/diagnose/tls_verification.go [hunk 7]
```diff
@@ -122,3 +123,30 @@ func TLSFileChecks(certFilePath, keyFilePath string) error {
...
+	if err != nil {
+		return fmt.Errorf("failed to read tls_client_ca_file: %w", err)
+	}
+
+	if !caPool.AppendCertsFromPEM(data) {
+		return fmt.Errorf("failed to parse CA certificate in tls_client_ca_file")
+	}
+	// TODO: should we run the CA cert through the TLSFileChecks function as well?
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification.go:139]: Yep, I'd like to add that to this PR if you don't mind?

**Comment by hghaf099** [comment]
This PR is old, and I will just close it. I have addressed the comments in https://github.com/hashicorp/vault/pull/11860/

**Pull request closed without merging**